extras/gizmo/gizmo.js

```diff
@@ -0,0 +1,325 @@
+     * @property {import('playcanvas').MeshInstance[]} meshInstances -
+     * array of mesh instances for rendering
+     */
+    /**
+     * The intersection data object.
+     *
+     * @type {IntersectData[]}
+     */
+    intersectData = [];
+
+    /**
+     * Creates a new Gizmo object.
+     *
+     * @param {import('playcanvas').AppBase} app - The application instance.
+     * @param {import('playcanvas').CameraComponent} camera - The camera component.
+     * @example
+     * const gizmo = new pcx.Gizmo(app, camera);
+     */
+    constructor(app, camera) {
+        super();
+
+        this.app = app;
+        this.camera = camera;
+
+        this._createLayer();
+        this._createGizmo();
+
+        this._updateScale();
+
+        this._onPointerDown = (e) => {
+            if (!this.gizmo.enabled || document.pointerLockElement) {
+                return;
+            }
+            const selection = this._getSelection(e.clientX, e.clientY);
+            if (selection[0]) {
+                e.preventDefault();
+            }
+            this.fire('pointer:down', e.clientX, e.clientY, selection[0]);
+        };
+        this._onPointerMove = (e) => {
+            if (!this.gizmo.enabled || document.pointerLockElement) {
+                return;
+            }
+            const selection = this._getSelection(e.clientX, e.clientY);
+            if (selection[0]) {
+                e.preventDefault();
+            }
+            this.fire('pointer:move', e.clientX, e.clientY, selection[0]);
+        };
+        this._onPointerUp = (e) => {
+            if (!this.gizmo.enabled || document.pointerLockElement) {
+                return;
+            }
+            this.fire('pointer:up');
+        };
+        this._onKeyDown = (e) => {
+            if (!this.gizmo.enabled) {
+                return;
+            }
+            this.fire('key:down', e.key, e.shiftKey, e.ctrlKey, e.metaKey);
+        };
+        this._onKeyUp = (e) => {
+            if (!this.gizmo.enabled) {
+                return;
+            }
+            this.fire('key:up');
+        };
+
+        app.on('update', () => this._updateScale());
+
+        app.on('destroy', () => this.detach());
+    }
+
+    set coordSpace(value) {
+        this._coordSpace = value ?? 'world';
+        this._updateRotation();
+    }
+
+    get coordSpace() {
+        return this._coordSpace;
+    }
+
+    set size(value) {
+        this._size = value;
+        this._updateScale();
+    }
+
+    get size() {
+        return this._size;
+    }
+
+    _getProjFrustumWidth() {
+        const gizmoPos = this.gizmo.getPosition();
+        const cameraPos = this.camera.entity.getPosition();
+        const dist = tmpV1.copy(gizmoPos).sub(cameraPos).dot(this.camera.entity.forward);
+        return dist * Math.tan(this.camera.fov * math.DEG_TO_RAD / 2);
+    }
+
+    _createLayer() {
+        const layerMap = this.app.scene.layers.layerIdMap;
+        if (layerMap.has(GIZMO_LAYER_ID)) {
```

Comment:
Don't hardcode GIZMO_LAYER_ID. Just test by the name if the layer exists, and leave whatever default ID a new layer gets.

But additionally, I'd remove the layer creation from the gizmo class completely .. a layer should be passed in in the constructor, giving user the flexibility to add it to any layer they already have.

Reply:
it should also not modify layers on the camera .. the user should do this.